dbms/src/DataStreams/MultiversionSortedBlockInputStream.h

```diff
@@ -0,0 +1,197 @@
+
+#include <deque>
+
+namespace DB
+{
+
+static const size_t MAX_ROWS_IN_MULTIVERSION_QUEUE = 8192;
+
+/* Deque with fixed memory size. Allows pushing gaps.
+ * frontGap() returns the number of gaps were inserted before front.
+ *
+ * Note: empty deque may have non-zero front gap.
+ */
+template <typename T>
+class FixedSizeDequeWithGaps
+{
+public:
+
+    struct ValueWithGap
+    {
+        size_t gap;
+        char value[sizeof(T)];
+    };
+
+    explicit FixedSizeDequeWithGaps(size_t size)
+    {
+        container.resize_fill(size + 1);
+    }
+
+    ~FixedSizeDequeWithGaps()
+    {
+        auto destruct_range = [this](size_t from, size_t to)
+        {
+            for (size_t i = from; i < to; ++i)
+                destructValue(i);
+        };
+
+        if (begin <= end)
+            destruct_range(begin, end);
+        else
+        {
+            destruct_range(0, end);
+            destruct_range(begin, container.size());
+        }
+    }
+
+    void pushBack(const T & value)
+    {
+        constructValue(end, value);
+        moveRight(end);
+        container[end].gap = 0;
+    }
+
+    void pushGap(size_t count) { container[end].gap += count; }
+
+    void popBack()
+    {
+        size_t curr_gap = container[end].gap;
+        moveLeft(end);
+        destructValue(end);
+        container[end].gap += curr_gap;
+    }
+
+    void popFront()
+    {
+        destructValue(begin);
+        moveRight(begin);
+    }
+
+    T & front() { return getValue(begin); }
+    const T & front() const { return getValue(begin); }
+
+    const T & back() const
+    {
+        size_t ps = end;
+        moveLeft(ps);
+        return getValue(ps);
+    }
+
+    size_t & frontGap() { return container[begin].gap; }
+    const size_t & frontGap() const { return container[begin].gap; }
+
+    size_t size() const
+    {
+        if (begin <= end)
+            return end - begin;
+        return end + (container.size() - begin);
+    }
+
+    bool empty() const { return begin == end; }
+
+private:
+    PODArray<ValueWithGap> container;
+
+    size_t gap_before_first = 0;
+    size_t begin = 0;
+    size_t end = 0;
+
+    void constructValue(size_t index, const T & value) { new (container[index].value) T(value); }
+    void destructValue(size_t index) { reinterpret_cast<T *>(container[index].value)->~T(); }
+
+    T & getValue(size_t index) { return *reinterpret_cast<T *>(container[index].value); }
+    const T & getValue(size_t index) const { return *reinterpret_cast<const T *>(container[index].value); }
+
+    void moveRight(size_t & index) const
+    {
+        ++index;
+
+        if (index == container.size())
+            index = 0;
+    }
+
+    void moveLeft(size_t & index) const
+    {
+        if (index == 0)
+            index = container.size();
+
+        --index;
+    }
+};
+
+class MultiversionSortedBlockInputStream : public MergingSortedBlockInputStream
+{
+public:
+    /// Don't need version column. It's in primary key.
```

Comment:
It is confusing. A version column could be not in the primary key. Do you mean it is in 'merging columns'?

Reply:
Added error with description "VersionedCollapsingMergeTree must have version column in primary key". Now, user must add it to primary key explicitly.